### ISSUE TITLE
HTTPProxy should overwrite IngressRoute if there is a conflict

### ISSUE THREAD

**Issue by youngnick** [projectcontour/contour#1949]
As part of #1689, I realised that we don't have a good upgrade path for IngressRoute to HTTPProxy right now.

Currently, AFAICT, having an IngressRoute and HTTPProxy with matching FQDNs and routes will render them both invalid.

This issue covers building something so that, if an IngressRoute and HTTPProxy overlap, then the HTTPProxy will take priority.

This allows us to have an upgrade process of:
- Have IngressRoutes deployed and working
- Build matching HTTPProxy objects (#1689 is to assist here)
- Apply new HTTPProxy objects
- Verify
- Remove either HTTPProxy objects (for rollback) or IngressRoute objects (for completion).

**Comment by davecheney** [comment]
Also, if a httpproxy takes priority over an ingressroute of the same hostname, the latter should be marked invalid, not orphaned.

**Comment by davecheney** [comment]
@youngnick I think this needs to land in 1.1 if we are to have a chance for people to use the translation tool with that version.

**Comment by youngnick** [comment]
This issue comes back to the problem that we've never really said explicitly what the priority policy is for the three types of config we support.

Currently, it works like this:
## Ingress
- Ingress objects are parsed into the DAG
## IngressRoute
- IngressRoute objects are parsed into the DAG.
- If there is a matching FQDN to one or more Ingresses, but different prefixes, the different routes are merged in to a single Envoy entry.
- If the FQDN and prefixes match, then IngressRoute config will overwrite the Ingress config.
The second item is required because Ingress objects are many-to-many with respect to FQDN; you can have many Ingresses that share an FQDN, and we need to work into this paradigm.

In all the above cases, the IngressRoute will be marked 'valid'.
## HTTPProxy

HTTPProxy objects are parsed into the DAG after Ingress and IngressRoute objects, *with the same semantics*

That is:
- If there is a matching FQDN to an Ingress or IngressRoute, but different prefixes (those are the only conditions that are present in the older forms), then everything is merged together.
- If there is a matching FQDN that also has a matching prefix, then the HTTPProxy will *overwrite* the Ingress or IngressRoute config.

In both of these cases, *both* the IngressRoute and the HTTPProxy will be marked as 'valid' in the DAG.

For ease of use, I propose that:
- IngressRoute import happens after HTTPProxy import into the DAG
- IngressRoute import will check to see if there is a matching FQDN in a root HTTPProxy. If so, it will be marked as 'invalid' with a reason of 'There's already a HTTPPRoxy matching fqdn.com' or similar.
- Any child IngressRoutes will be left orphaned and thus be invalid.

This is a change in behaviour that will allow users to see that their IngressRoute config is being correctly superceded by new HTTPProxy config, and that removing the HTTPProxy config (by deleting the objects) will restore the IngressRoute config.

This will allow for a more graceful conversion process from IngressRoute to HTTPProxy.

**Comment by jpeach** [comment]
> If so, it will be marked as 'invalid' with a reason of 'There's already a HTTPPRoxy matching fqdn.com' or similar.

Can you include an object reference to the proxy that took precedence?

**Comment by davecheney** [comment]
> Can you include an object reference to the proxy that took precedence?

It is possible, but it is more work as we'd have to keep a table of FQDN to HTTPProxy mappings. My suggestion to nick was to not do this at this time. If needed we can come back and add that, but I suspect it won't be needed as this feature only has a limited lifespan.

**Comment by davecheney** [comment]
@youngnick I think the discussion would be clearer if the policy were separated from the implementation -- don't worry about the dag, the builder, status, or any of the other guts inside Contour, they're implementation details. 

From the policy point of view we have virtualhosts, they have one or more routes, and are each identified by a fully qualified domain name. I think the key policy question is: _which object is permitted to define the routes for a virtualhost_. 

For k8s ingress, it's unknown. Multiple virtualhosts may be mentioned per ingress object, and multiple ingress objects may define routes on a single virtual host. The only way to approach this is just say once we've finished evaluating the set of ingress objects visible to Contour, there will be a (possibly empty) set of virtualhosts defined with some routes on them.

For no good reason this logic has historically be extended to HTTPProxy and IngressRoute objects -- the net effect is after evaluating all three sets of objects, we have some virtualhosts with some routes defined on them.

There are a few policies in place, inside the set of HTTPProxy roots, a virtualhost may be defined only once. The same logic applies to IngressRoute, however it does not apply to the union of  HTTPPRoxy and IngressRoutes roots.

Because of this oversight we have a more complicated problem of ordering. routes defined on virtualhosts in Ingress objects are overplayed with routes from IngressRoute objects which are in turn overlayed with routes defined in HTTPProxy objects. If we're lucky these route sets are disjoint and the order of evaluating the different object sets doesn't matter. If our luck ran out then we have users in the first who have deployed all three and have overlapping route sets; specifically prefix routes because that is all Ingress and IngressRoute can define. If we reverse the order of evaluation of HTTPProxy objects and IngressRoute objects so the former can "claim" a virtualhost then Contour 1.1 users will experience a different routing set.

The situation where users have routes defined on a single virtualhost across a HTTPProxy and IngressRoute object if we change policy to say that a virtualhost can only be defined across the union of those two object sets then the routes defined on the IngressRoute will removed. This might be too far of a bridge for some users.

Perhaps the answer is to keep the existing logic exactly as it is.

**Comment by youngnick** [comment]
That was my original idea, but I was worried about the user experience of being able to tell if your IngressRoute is being overwritten by a HTTPProxy.

I think that it's probably okay, and will certainly make it much easier to land this (basically, merge the tests in the associated PR and it's done).

**Comment by davecheney** [comment]
> That was my original idea, but I was worried about the user experience of being able to tell if your IngressRoute is being overwritten by a HTTPProxy.

I think the problem is we were thinking HTTPProxy x completely occludes IngressRoute x, but the case where there is partial overlap means we probably shouldn't change the order in which IngressRoute and HTTPProxy objects are ordered, nor can we change the interpretation to say if IngressRoute and HTTPProxy share a virtualhost, the former is invalidated.

**Comment by davecheney** [comment]
@jpeach @stevesloka WDYT?

**Comment by davecheney** [comment]
Don't forget the fallback postion is always;

1. k get -o yaml 
2. translate
3. k delete ir
4. k apply httpproxy

The window for downtime is on the order of seconds, most of which is kubectl latency. Don't let perfection be the enemy of the good.

**Comment by jpeach** [comment]
So the risk here is that you can end up with your production configuration depending on both HTTPProxy and IngressRoute. In principle at least, they could both be changing at the same time.

I'm tempted to suggest this could be made marginally safer by a Contour config that you could toggle to ignore IngressRoute, but that seems very similar to switching Dave's steps (3) and (4) above, and I expect that large clusters may need to migrate service at a time.

> The situation where users have routes defined on a single virtualhost across a HTTPProxy and IngressRoute object if we change policy to say that a virtualhost can only be defined across the union of those two object sets then the routes defined on the IngressRoute will removed. This might be too far of a bridge for some users.

So, the argument here is that if you are in this inconsistent state, the tooling will bring you out of it, so everything will be OK in the end? That sounds right to me, as long as the objects aren't mutating too rapidly. Is it possible for the tooling to deterministically know the types are in sync?

**Comment by youngnick** [comment]
The tooling is a very simple translation. So it will take an IngressRoute and output a matching HTTPProxy. In standard use, this issue should not come up, and the process is:
- Have a working IngressRoute
- Back up the IngressRoute
- Run ir2proxy
- Check the generated HTTPProxy
- Apply the generated HTTPProxy
- Check your service is working
- Delete the IngressRoute
- Check your service is working

If the service is not working after you first apply the HTTPProxy, then you can remove it to revert to IngressRoute.

The issue came up more because writing tests for the current behaviour showed the ways that this could possibly go wrong.

To be honest, I think we've all spent enough cycles on this now, so my current plan is:
- merge the tests to cover the current behaviour (#1981). That will close out this issue and unblock 1.1
- finish ir2proxy as soon as possible.

**Comment by youngnick** [comment]
To summarise this further one more time:
- The tests in #1981 lock in the current behaviour of IngressRoute being superseded by HTTPProxy.
- We won't do anything with adding status, as it doesn't seem possible to do it without a lot of work that will be obsolete once we remove IngressRoute anyway.

@jpeach @stevesloka what do you think?

**Comment by jpeach** [comment]
Yeh, that sounds fine :)
